scipy/stats/_binned_statistic.py

```diff
@@ -584,7 +584,13 @@ def binned_statistic_dd(sample, values, statistic='mean',
             result[vv, a] = flatsum[a] / flatcount[a]
     elif statistic == 'std':
         result.fill(0)
-        _calc_binned_statistic(Vdim, binnumbers, result, values, np.std)
+        flatcount = np.bincount(binnumbers, None)
+        a = flatcount.nonzero()
+        for vv in builtins.range(Vdim):
+            flatsum = np.bincount(binnumbers, values[vv])
+            delta = values[vv] - flatsum[binnumbers] / flatcount[binnumbers]
```

Comment:
The original code in `_calc_binned_statistic` checks explicitly if there's at least 2 data points in the bin, and otherwise skips the calculation. The code here does look correct, however ot looks like there is no test that covers this case. Could you have a look at that @teddyort, adding a test with a few empty bins?

Reply:
@rgommers I looked into this, and agree that the current PR will not change the current behavior in the special case of < 2 data points per cell.

However, while preparing a unit test, I realized you may have uncovered a bug in the current implementation.

If we take the current std unit test: https://github.com/scipy/scipy/blob/e4029e66af17383697c34a519f8a46e52b08eb23/scipy/stats/tests/test_binned_statistic.py#L96-L104 

and simply set bins=1000 (this ensures many empty bins because x,v are 100 random numbers) the test fails currently. The reason is that `np.std` gives `nan` for empty bins, while currently `binned_statistic` gives 0 with `'std'`.

In my opinion, the numpy result is correct here since `'std'` is not defined for an empty cell. It also fits better with the other statistics min/max/median which all give `nan`. It seems (based on the comment in the code) that the reason this check was implemented was to improve the performance by skipping any bins with < 2 values to save time. This PR would negate that need as we no longer call `np.std` in a loop. Therefore, I would advocate we remove this check, and make the output of `'std'` match `np.std`. 

However, this would not be a backwards compatible change. If we leave it as is, the output will remain unchanged, but perhaps this strange behavior should be documented.

Any thoughts?

Reply:
Good catch. I agree, the result should be `nan` here. I'd say this is a bug, so we don't have to worry about backwards compatibility - we just document the change in the release notes.

Reply:
Sounds good. I've changed the empty bin default to `nan` and added a test.